components/core/src/config/templatehostidextractor.ts

```diff
@@ -0,0 +1,60 @@
+/**
+ * Copyright 2023 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {PluggableFactory} from '../integrations/pluggablefactory';
+import {JsonMap} from '../util/jsonvalue';
+import {TenantIdExtractor} from './tenantidextractor';
+
+export interface TemplateHostIdExtractorOptionsJson extends JsonMap {
+  urlTemplate: string;
+}
+
+export class TemplateHostIdExtractor implements TenantIdExtractor {
+  readonly type = 'tenantIdExtractor';
+
+  private urlTemplate: string;
+  private idExtractRegexp: RegExp;
+
+  constructor(options: TemplateHostIdExtractorOptionsJson) {
+    this.urlTemplate = options.urlTemplate;
+    this.idExtractRegexp = new RegExp(
+      this.urlTemplate.replace('$', '([a-z0-9_-]+)')
+    );
+  }
+
+  getTenantId(reqUrl: string): string | undefined {
+    const result = this.idExtractRegexp.exec(reqUrl);
+    if (result && result.length > 0 && result[1]) {
+      return result[1];
+    }
+    return undefined;
```

Comment:
I decided not to log this, because there are lots of valid requests to a server that return undefined here. Any request to a server-wide custom endpoint, for example, is first checked to see if it's a TenantNode request.